Include/pyport.h

```diff
@@ -184,8 +184,23 @@ typedef Py_ssize_t Py_ssize_clean_t;
 #  define Py_LOCAL_INLINE(type) static inline type
 #endif
 
+// Preprocessor check for a builtin preprocessor function.
+// Return 0 if the __has_builtin() macro is not defined.
+//
+// __has_builtin() is available in Clang and in GCC 10+.
+#ifdef __has_builtin
+#  define _Py__has_builtin(x) __has_builtin(x)
+#else
+#  define _Py__has_builtin(x) 0
+#endif
+
+// Avoid calls to external functions when possible,
+#if _Py__has_builtin(__builtin_memcpy_inline) || defined(__clang__)
+#  define Py_MEMCPY __builtin_memcpy_inline
+#elif _Py__has_builtin(__builtin_memcpy) || defined(__GNUC__)
+#  define Py_MEMCPY __builtin_memcpy
+#else
 // bpo-28126: Py_MEMCPY is kept for backwards compatibility,
```

Comment:
It seems previously we were (silently) deprecating using `Py_MEMCPY`, in favor of just calling `memcpy`, since the compilers all seem to optimize it anyways. (See e.g. [write_u32](https://github.com/python/cpython/blob/24facd60f6e5458cc783e429fe63d6a80d476f84/Include/internal/pycore_code.h#L285) and friends in pycore_code.h, which is definitely required to be fast, as it is part of the bytecode interpreter inline caching infrastructure.) Are we now reversing our position on this topic? If so, this comment is outdated. Otherwise, maybe just write `memcpy` and be happy?

Reply:
> Otherwise, maybe just write memcpy and be happy?

Ideally, I'd be using memcpy only! There's a small problem, using memcpy naively in this function causes Py_INCREF to not get inlined. However, I managed to get around this by using the builtin functions (`__builtin_memcpy_inline`, `__builtin_memcpy `) instead. 

As soon as we change the builtin functions with plain `memcpy`, we get:
```
> nm ./python | grep "Py_INCREF"
00000000001d8340 t Py_INCREF.lto_priv.0
00000000001d8340 t Py_INCREF.lto_priv.1
...
```

Unfortunately this does have a slight performance hit (which we also noticed it when running Instagram, as we were originally just using `memcpy`). Getting some quick numbers with Valgrind and looking at the disassembly we now get::

Plain `memcpy`:
```
0000000000199b16 <_Py_IncRef>:
  199b16:       f3 0f 1e fa             endbr64
  199b1a:       e9 f1 e8 03 00          jmpq   1d8410 <Py_INCREF.lto_priv.0>
 00000000001d8410 <Py_INCREF.lto_priv.0>:
  1d8410:       8b 07                   mov    (%rdi),%eax
  1d8412:       83 c0 01                add    $0x1,%eax
  1d8415:       72 02                   jb     1d8419 <Py_INCREF.lto_priv.0+0x9>
  1d8417:       89 07                   mov    %eax,(%rdi)
  1d8419:       c3                      retq
  1d841a:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)

Richards Benchmark:
  I   refs:      516,046,487
  I1  misses:        405,352
```

Builtins functions (i.e: `__builtin_memcpy`):
```
000000000019c2f9 <_Py_IncRef>:
  19c2f9:       f3 0f 1e fa             endbr64
  19c2fd:       8b 07                   mov    (%rdi),%eax
  19c2ff:       83 c0 01                add    $0x1,%eax
  19c302:       72 02                   jb     19c306 <_Py_IncRef+0xd>
  19c304:       89 07                   mov    %eax,(%rdi)
  19c306:       c3                      retq

Richards Benchmark:
  I   refs:      513,502,941
  I1  misses:        400,129
```

Thoughts? cc @markshannon @carljm @ericsnowcurrently

Reply:
Ignore this message, I managed to get this working using a union in the PyObject base structure to directly address the lower 32 bits. No need for a memcpy anymore